### DISCUSSION TITLE
"spin up" should build when needed

### DISCUSSION THREAD

**Discussion by tshepang** [spinframework/spin#692] [Ideas]
I forgot to first `spin build`, so I had this...
```
❯ spin up
Error: Cannot create module for component foo from file /home/tshepang/spin/foo/target/wasm32-wasi/release/foo.wasm

Caused by:
    0: failed to read input file
    1: No such file or directory (os error 2)
Error: exit status: 1
```
Not opening issue, for there could be a reason this does not happen automatically.

**Comment by radu-matei** [comment]
The main reason we don't currently have this implemented is because reliably understanding "when needed" is a bit difficult.

If the module source is not present, this would be straightforward, and we could run `spin build`. 
But what should happen if the module _is_ present? We could hash the entire directory and if anything changed, we would rebuild (we are exploring something similar for `spin deploy`, cc @michelleN)

Is that what you had in mind?

Thanks!

**Reply by tshepang** [reply]
One special case would be automatically building if `.wasm` is found to not exist.

Regarding hashing things, I was thinking that `spin up` could behave like `cargo watch`, and just rebuild when changes are detected (probably using something inotify). It could also have an option to disable that watcher behavior, in case some do not desire the magic.
